src/bloqade/pyqrack/squin/runtime.py

```diff
@@ -0,0 +1,323 @@
+from typing import Any
+from dataclasses import dataclass
+
+import numpy as np
+from kirin.dialects import ilist
+
+from pyqrack.pauli import Pauli
+from bloqade.pyqrack import PyQrackQubit
+
+
+@dataclass(frozen=True)
+class OperatorRuntimeABC:
+    def apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        raise NotImplementedError(
+            "Operator runtime base class should not be called directly, override the method"
+        )
+
+    def control_apply(self, *qubits: PyQrackQubit, adjoint: bool = False) -> None:
+        raise RuntimeError(f"Can't apply controlled version of {self}")
+
+    def broadcast_apply(self, qubits: ilist.IList[PyQrackQubit, Any], **kwargs) -> None:
+        for qbit in qubits:
+            if not qbit.is_active():
+                continue
+
+            self.apply(qbit, **kwargs)
```

Comment:
What if the operator accepts two qubits? you should batch the qubits based on the operator size.